### PR TITLE
Add Jax Tensor Support

### PR DESCRIPTION
**Pull request by nbardy**
TODO:
~Add tests~
Research more about numpy tensor and what other types we might want to support.

### CODE REVIEW

File: wandb/util.py [hunk 1]
```diff
@@ -190,9 +190,13 @@ def is_tf_tensor(obj):
     return isinstance(obj, tensorflow.Tensor)
 
 
+def is_jax_tensor_typename(typename):
+    return typename.endswith('_FilledConstant') or typename.endswith('DeviceArray')
```
Comment (vanpelt) [inline wandb/util.py]: Don't like the endswith check here.  There could be other objects that end with these strings.  I assume the types start with "jax"?

File: wandb/util.py [hunk 2]
```diff
@@ -248,6 +252,9 @@ def json_friendly(obj):
 
     if is_tf_tensor_typename(typename):
         obj = obj.eval()
+    elif is_jax_tensor_typename(typename):
+        import numpy
```
Comment (vanpelt) [inline wandb/util.py]: If we're sure it's jax we should `import jax.numpy as np` the user might not have numpy.

Comment (nbardy) [reply wandb/util.py]: This is the recommended way to move the tensor from the gpu to cpu in jax. Jax support will have to include numpy support.

**Comment by nbardy** [comment]
Alright I've done some more research. Jax is built on top of numpy and uses it internally, and numpy is the recommended way to save jax arrays[0].


[0] https://github.com/google/jax/blob/master/jax/numpy/lax_numpy.py#L148

This PR adds support for logging some of the internal jax types. If you log try to log a tensor inside a jax function that has been transformed via `jit` or `grad` you end up with values that can be lazy, and when we try to write them later they could be garbage collected. Casting to a `numpy` array freezes a copy in memory. Which allows us to safely log it at any time.

The only two I have needed in my testing so far are `jax.interpreters.xla.DeviceArray`

I need to look around the Jax Internals more and see what other types we would need to cover before I feel comfortable claiming "Jax Support" The only two that I've bumped into are `jax.interpreters.xla.DeviceArray`

**Comment by coolalexzb** [comment]
Hey guys. My name is Bo Zheng. I want to help contribute to this PR. Is this PR still needed? Besides, can anyone provide developer documentation? I think the documentation I can find on `wandb` website is just for users. Thank you! @nbardy @vanpelt

**Comment by nbardy** [comment]
@coolalexzb I haven't been using JAX myself lately and don't know a lot about the  library. I would love to see someone else pick up the torch and bring the integration to completion. You can start here or start from a new branch.

Our documentation repo is here:
https://github.com/wandb/gitbook

Any contributions there are welcome.

**Comment by coolalexzb** [comment]
Thank you! I will go ahead. @nbardy

**Comment by coolalexzb** [comment]
Hi! I have some questions about JAX support.

1. I find a JAX example [here](https://github.com/wandb/examples/blob/master/jax-mnist/train.py). Does `wandb` support JAX now? 
2. What are the existing problems in this PR? Can you explain the work that is not completed in this PR?
3.  I find it difficult to trace back how `wandb` works and contribute to add JAX support. Can you give some suggestions? 

Thank you! @nbardy

**Comment by nbardy** [comment]
@coolalexzb Wandb is framework agnostic so you can use it with any library, but we have advanced integrations to offer better functionality with specific frameworks.

This PR allows you to log jax tensor types(or tensor like) as histograms via wandb.

Finding all the native JAX types and adding them to the type checks is the remaining work for this PR.

**Comment by dmitryduev** [comment]
We do have `jax` tensor support now: https://github.com/wandb/client/blob/2c5c13727b26b29d485ed8b7fbf293154e2c7cd9/wandb/util.py#L514.
@raubitsj @vanpelt can I close this PR?

**Pull request closed without merging**